### PR TITLE
Fix Two Snapshot Clone State Machine Bugs

### PR DESCRIPTION
**Pull request by original-brownbear**
There are two separate but closely related bug fixes in this PR:
1. When two snapshot clones would initialize concurrently we could get into a state
where one is in front of the other in the queue snapshots array but its shard states
are in fact queued behind the other snapshot or clone. Tightly linked to this,
snapshot cloning would not account for snapshot deletes when queuing shard snapshots
which could lead to races where both delete and clone are running concurrently for a shard.
2. As a result of fixing the first issue and writing a test for it, it also became obvious
that a finished delete was not properly accounted for when it comes to starting snapshot
clones that could now queue behind a delete.

The first bug can be seen in https://gradle-enterprise.elastic.co/s/iii3ntyxbil5i for example, the random blocking on reading metadata is just there to make it highly likely to occur. Unfortunately, a 100% reproducer for this wasn't possible with the current test infrastructure but I think this combined with the very strict asserts is good enough for now and tests can be enhanced when resolving TODOs added by this PR.

A few notes:

* I added TODOs about drying things up instead of going for the perfect solution in spots right away to keep this PR at least somewhat digestible 
* A large part of the changes in this PR are just enhancements to asserts (that would have caught the bugs in this right away) and related infrastructure

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -476,25 +476,50 @@ public ClusterState execute(ClusterState currentState) {
...
                         updatedEntry = cloneEntry.withClones(clonesBuilder.build());
-                        updatedEntries.set(i, updatedEntry);
+                        if (queuedShards) {
+                            // We queued up some shards based on the in-flight operations found in all snapshots for the current
+                            // repository, so in order to make sure we don't set a shard to QUEUED before (as in before it in the
+                            // `updatedEntries` list) one that is actively executing we just put it to the back of the list as if we had
+                            // just created the entry
+                            // TODO: If we could eventually drop the snapshot clone init phase we don't need this any longer
+                            updatedEntries.remove(i);
+                            updatedEntries.add(updatedEntry);
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:519]: This is the only place where the clone is re-queued right? I was wondering if this could lead to starvation in some really bad scenarios, but I think that's not possible?

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java [hunk 2]
```diff
@@ -471,6 +472,25 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                 }
             }
             builder.endArray();
+            if (isClone()) {
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:475]: Is this related to this pr?

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/CloneSnapshotIT.java [hunk 3]
```diff
@@ -571,6 +575,40 @@ public void testStartCloneWithSuccessfulShardSnapshotPendingFinalization() throw
...
+        final ActionFuture<AcknowledgedResponse> deleteFuture = startDeleteSnapshot(repoName, randomFrom(snapshotNames));
+        waitForBlock(masterName, repoName);
+        awaitNDeletionsInProgress(1);
+
+        final ActionFuture<AcknowledgedResponse> cloneFuture = startClone(repoName, sourceSnapshot, "target-snapshot", indexName);
+        logger.info("--> waiting for snapshot clone to be fully initialized");
+        awaitClusterState(state -> {
+            for (SnapshotsInProgress.Entry entry : state.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY).entries()) {
+                if (entry.clones().isEmpty() == false) {
+                    assertEquals(sourceSnapshot, entry.source().getName());
```
Comment (fcofdez) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/CloneSnapshotIT.java:601]: Maybe we could assert here that the shard snapshots are queued?

**Review by fcofdez — APPROVED** [review]
Other than the minor comments it LGTM but I'll wait for @tlrx opinion on the PR, it's possible that I'm missing something.

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/CloneSnapshotIT.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/CloneSnapshotIT.java:602]: ++ I added that now. Technically this is ensured by assertions we have on the cluster state in `SnapshotsService` but practically this makes the test a lot clearer in its intent :)

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:475]: Sort of, it's of necessary to make assertion failure messages useful, that's where I noticed this was missing.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:519]: > This is the only place where the clone is re-queued right

Right, we're only doing this here.

> I was wondering if this could lead to starvation in some really bad scenarios, but I think that's not possible?

I think we're good here even in theory. We do all the shard count fetching on the `SNAPSHOT` pool and any operation that would keep adding more and more operations to the list so quickly that we never get our turn would have to do some work on the `SNAPSHOT` pool eventually so we're safe on that front :)

**Comment by original-brownbear** [comment]
Thanks @fcofdez ! And ++ we definitely need @tlrx :eyes: on this as well (sorry in advance Tanguy! :))

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java (hunk 2 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:475]: That looks strange that `Entry` implements `ToXContent` just for rendering assertion error messages? I guess it's fine but I find it strange (and it should implement `ToXContentObject`)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 4]
```diff
@@ -476,25 +476,50 @@ public ClusterState execute(ClusterState currentState) {
...
+                    if (cloneEntry.repository().equals(entry.repository()) == false) {
+                        // different repo => just continue without modification
+                        continue;
+                    }
+                    if (cloneEntry.snapshot().getSnapshotId().equals(entry.snapshot().getSnapshotId())) {
                         final ImmutableOpenMap.Builder<RepositoryShardId, ShardSnapshotStatus> clonesBuilder =
                                 ImmutableOpenMap.builder();
-                        final InFlightShardSnapshotStates inFlightShardStates =
-                            InFlightShardSnapshotStates.forRepo(repoName, snapshotsInProgress.entries());
+                        final boolean readyToExecute = currentState.custom(
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:487]: `readyToExecute` kind of lack of meaning, maybe `startedSnapshotDeletion` (+ `anyMatch`)?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 5]
```diff
@@ -797,11 +822,15 @@ private static boolean assertNoDanglingSnapshots(ClusterState state) {
...
-                for (ObjectCursor<ShardSnapshotStatus> value : entry.shards().values()) {
+                for (ObjectCursor<ShardSnapshotStatus> value : (entry.isClone() ? entry.clones() : entry.shards()).values()) {
                     if (value.value.equals(ShardSnapshotStatus.UNASSIGNED_QUEUED)) {
                         assert reposWithRunningDelete.contains(entry.repository())
                                 : "Found shard snapshot waiting to be assigned in [" + entry +
                                 "] but it is not blocked by any running delete";
+                    } else if (value.value.isActive()) {
+                        assert reposWithRunningDelete.contains(entry.repository()) == false
+                                : "Found shard snapshot actively executing in [" + entry +
+                                "] when it should be blocked by a running delete";
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:833]: :+1:

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 6]
```diff
@@ -1929,50 +1961,89 @@ private SnapshotsInProgress updatedSnapshotsInProgress(ClusterState currentState
...
-                        // Collect waiting shards that in entry that we can assign now that we are done with the deletion
-                        final List<ShardId> canBeUpdated = new ArrayList<>();
-                        for (ObjectObjectCursor<ShardId, ShardSnapshotStatus> value : entry.shards()) {
-                            if (value.value.equals(ShardSnapshotStatus.UNASSIGNED_QUEUED)
-                                    && reassignedShardIds.contains(value.key) == false) {
-                                canBeUpdated.add(value.key);
+                        // TODO: dry up redundant computation and code between clone and non-clone case, in particular reuse
+                        //  `inFlightShardStates` across both clone and standard snapshot code
+                        if (entry.isClone()) {
+                            // Collect waiting shards that in entry that we can assign now that we are done with the deletion
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1979]:
```suggestion
                            // Collect waiting shards from that entry that we can assign now that we are done with the deletion
```

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 7]
```diff
@@ -1929,50 +1961,89 @@ private SnapshotsInProgress updatedSnapshotsInProgress(ClusterState currentState
...
+                                if (value.value.equals(ShardSnapshotStatus.UNASSIGNED_QUEUED)
+                                        && alreadyReassigned(value.key.indexName(), value.key.shardId(), reassignedShardIds) == false) {
+                                    canBeUpdated.add(value.key);
+                                }
+                            }
+                            // TODO: the below logic is very similar to that in #startCloning and both could be dried up against each other
+                            //       also the code for standard snapshots could make use of this breakout as well
+                            final boolean readyToExecute = updatedDeletions.getEntries().stream().noneMatch(
+                                    e -> e.repository().equals(repoName) && e.state() == SnapshotDeletionsInProgress.State.STARTED);
+                            if (readyToExecute == false || canBeUpdated.isEmpty()) {
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1991]: Maybe evaluate `canBeUpdated.isEmpty()` first and then only compute `readyToExecute` if needed?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 8]
```diff
@@ -1987,6 +2058,15 @@ private SnapshotsInProgress updatedSnapshotsInProgress(ClusterState currentState
             }
             return changed ? SnapshotsInProgress.of(snapshotEntries) : null;
         }
+
+        private void markShardReassigned(String indexName, int shardId, Map<String, Set<Integer>> reassignments) {
+            final boolean added = reassignments.computeIfAbsent(indexName, k -> new HashSet<>()).add(shardId);
+            assert added;
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2064]: Maybe add some message in case it occurs in tests?

**Review by tlrx — APPROVED** [review]
LGTM - Logic looks good to me, I left only minor comments (feel free to apply or not).

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 7 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1991]: ++

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 4 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:487]: ++ this is 100% the better name, I'll leave it as is for a sec though so it's less confusing to review when drying things up (we use the same mechanics and var name in the normal snapshot shard assignments path ... and it's equally weird there in hindsight :))

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 8 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: ++

**Comment by original-brownbear** [comment]
Thanks Francisco and Tanguy!
